### PR TITLE
HierarchicalStructure set avtiveItems when renderHierarchyAction with remark of Davma

### PR DESCRIPTION
**Pull request by IsaMic**

### CODE REVIEW

File: Controller/EmbedController.php [hunk 1]
```diff
@@ -19,7 +20,12 @@ public function __construct(ClientRequestManager $manager)
 
     public function renderHierarchyAction(string $template, string $parent, string $field, int $depth = null, array $sourceFields = [], array $args = []): Response
     {
-        $hierarchy = $this->clientRequest->getHierarchy($parent, $field, $depth, $sourceFields);
+		$hierarchy = $this->clientRequest->getHierarchy($parent, $field, $depth, $sourceFields);
+		
+		//Search all active items if given emsLink of currentPae
+		if(isset($args['emsLink']) && $args['emsLink'] instanceof EMSLink) {
+		    $hierarchy->setAllActives($args['emsLink']->getOuuid());
+		}
```
Comment (dameert) [inline Controller/EmbedController.php]: Remove the comment and rename the key 'emsLink' to 'currentPage'.
Your code $args['currentPage'] instanceof EMSLink already explains that it is an EMSlink and for the currentpage (or maybe 'activePage' is better?

**Comment by dameert** [comment]
In general I think it is strange to set the active item in the renderFunction, can't we do this in the getHierarchy function on a recursive way?

In the loop of getHierarchy pass the args. If the current page is the element in your hierarchy, set it as active. The loop will go up in the recurstion, so after creation of the child, we can ask if the child is active. If true, then we are also active.

File: Helper/Elasticsearch/HierarchicalStructure.php [hunk 2]
```diff
@@ -59,6 +61,39 @@ public function setData($data): HierarchicalStructure
...
+        $this->active = $active;
+
+        return $this;
+    }
+    /**
+     * Search all active items for a given Id
+     * @return HierarchicalStructure
+     */
+    public function setAllActives(string $currentId): HierarchicalStructure
+    {
```
Comment (dameert) [inline Helper/Elasticsearch/HierarchicalStructure.php]: This function can be removed if we use the recursion of getHierarchy

File: Controller/EmbedController.php (hunk 1 above)
Comment (Davidmattei) [reply Controller/EmbedController.php]: I prefer 'activeChild'

Comment (IsaMic) [reply Controller/EmbedController.php]: I changed the name 3 times because I hesitated with 'currentPage'..
'activePage' is better I choose it

Comment (IsaMic) [reply Controller/EmbedController.php]: ok for activeChild

File: Helper/Elasticsearch/HierarchicalStructure.php (hunk 2 above)
Comment (IsaMic) [reply Helper/Elasticsearch/HierarchicalStructure.php]: done

File: Controller/EmbedController.php (hunk 1 above)
Comment (IsaMic) [reply Controller/EmbedController.php]: done

File: Controller/EmbedController.php [hunk 3]
```diff
@@ -19,7 +20,11 @@ public function __construct(ClientRequestManager $manager)
 
     public function renderHierarchyAction(string $template, string $parent, string $field, int $depth = null, array $sourceFields = [], array $args = []): Response
     {
-        $hierarchy = $this->clientRequest->getHierarchy($parent, $field, $depth, $sourceFields);
+        $activeChild = '';
+        if(isset($args['activeChild']) && $args['activeChild'] instanceof EMSLink) {
+            $activeChild = $args['activeChild']->getOuuid();
+        }
+        $hierarchy = $this->clientRequest->getHierarchy($parent, $field, $depth, $sourceFields, $activeChild);
```
Comment (dameert) [inline Controller/EmbedController.php]: renderHierarchy is called from a twig function. The twig function will send an Ouuid. (we cannot create an EMSLink object in twig).

so $args['activeChild'] will be a simple string, that is in fact the Ouuid. (So your code will be much simpler.):

$hierarchy = $this->clientRequest->getHierarchy($parent, $field, $depth, $sourceFields, $args['activeChild'] ?? null);

File: Helper/Elasticsearch/ClientRequest.php [hunk 4]
```diff
@@ -257,13 +257,13 @@ public function getHierarchy(string $emsKey, string $childrenField, int $depth =
             return null;
         }
 
-        $out = new HierarchicalStructure($item['_type'], $item['_id'], $item['_source']);
+        $out = new HierarchicalStructure($item['_type'], $item['_id'], $item['_source'], ($item['_id'] === $activeChild ? true : false));
```
Comment (dameert) [inline Helper/Elasticsearch/ClientRequest.php]: Pass the activeChild as parameter, and do the logic and "setActive" in the constructor

File: Helper/Elasticsearch/HierarchicalStructure.php [hunk 5]
```diff
@@ -14,13 +14,16 @@ class HierarchicalStructure
     private $source;
     /** @var mixed  */
     private $data;
+    /** @var bool  */
+    private $active = false;
 
-    public function __construct(string $type, string $id, array $source)
+    public function __construct(string $type, string $id, array $source, bool $active)
```
Comment (dameert) [inline Helper/Elasticsearch/HierarchicalStructure.php]: string $activeId

File: Resources/doc/twig.md [hunk 6]
```diff
@@ -47,19 +47,19 @@ and a Symfony\Component\Finder\SplFileInfo object as value.
     'field': 'children',
     'depth': 5,
     'sourceFields': [],
-    'args': {'extra': 'test'}
+    'args': {'activeChild': emslink, 'extra': 'test'}
```
Comment (dameert) [inline Resources/doc/twig.md]: emslink: add example ouuid

Comment (IsaMic) [reply Resources/doc/twig.md]: emslink is the value.

**Pull request merged**